services/builder/domain/usecases/workflow_helpers.go

```diff
@@ -76,6 +76,76 @@ func (u *Usecases) buildCluster(ctx *builder.Context) (*builder.Context, error)
+		// The proxy mode is off in the current state. The proxy modes is on in the desired state.
+		// The proxy mode is on in the current state. The desired state has at least one Hetzner node and the proxy mode is default.
+		// The proxy mode is on in the current state. The desired state doesn't have any Hetzner nodes and the proxy mode is default.
+		// The proxy mode is off in the current state. The proxy mode is on in the desired state.
+		// The proxy mode is off in the current state. The desired state has at least one Hetzner node and the proxy mode is default.
+
+		// The following use cases represents scenarios when the proxy envs don't have to be updated:
+		// The proxy mode is off in both current and the desired state.
+		// The proxy mode is default in both current and the desired state. Both states don't have any Hetzner nodes.
+		// The proxy mode is default in the current state. The current state doesn't have any Hetzner nodes. The proxy is turned off in the desired state.
+		// The proxy mode is off in the current state. The desired state doesn't have any Hetzner nodes and the proxy mode is default.
+
+		if currProxySettings.Mode == offMode && desiredProxySettings.Mode == offMode {
+			// The proxy is and was turned off in both cases.
+			updateProxyEnvsFlag = false
+		} else if currProxySettings.Mode == defaultMode && desiredProxySettings.Mode == defaultMode &&
+			!builder.HasHetznerNode(ctx.CurrentCluster.ClusterInfo) && !builder.HasHetznerNode(ctx.DesiredCluster.ClusterInfo) {
+			// The proxy is in default mode without Hetzner node in both cases.
+			updateProxyEnvsFlag = false
+		} else if currProxySettings.Mode == defaultMode && !builder.HasHetznerNode(ctx.CurrentCluster.ClusterInfo) &&
+			desiredProxySettings.Mode == offMode {
+			// The proxy was in default mode without Hetzner node and is turned off.
+			updateProxyEnvsFlag = false
+		} else if currProxySettings.Mode == offMode && desiredProxySettings.Mode == defaultMode &&
+			!builder.HasHetznerNode(ctx.DesiredCluster.ClusterInfo) {
+			// The proxy was in off mode. Now it is in default mode without Hetzner node.
+			updateProxyEnvsFlag = false
+		}
+	} else if ctx.DesiredCluster != nil {
+		// The cluster wasn't build yet because currentState is nil
+		// but we have to check if the proxy is turned on or in a default mode with Hetzner node in the desired state.
+		desiredProxySettings := ctx.DesiredCluster.InstallationProxy
+
+		if desiredProxySettings.Mode == defaultMode && builder.HasHetznerNode(ctx.DesiredCluster.ClusterInfo) {
+			updateProxyEnvsFlag = true
+		} else if desiredProxySettings.Mode == "on" {
+			updateProxyEnvsFlag = true
+		}
+	}
+
+	// HttProxyUrl and NoProxyList will be set before first task in ansibler and then updated after ansibler Install VPN phase.
+	ctx.ProxyEnvs = &spec.ProxyEnvs{
+		UpdateProxyEnvsFlag: updateProxyEnvsFlag,
+	}
+
```

Comment:
move this into `internal/builder_context` file into its own function

```go
func (c *BuilderContext) DetermineProxyUpdate() error {
....
}
```


Further, rewrite the if else if chain into a switch that returns an error in the default case. By returning an error the build will not progress if we cannot determine if an proxy update is needed, avoiding failure.

```go
hetznerInDesired := builder.HasHetznerNode(ctx.DesiredCluster.ClusterInfo)
hetznerInCurrent := builder.HasHetznerNode(ctx.CurrentCluster.ClusterInfo)
switch {
case currProxySettings.Mode == offMode && desiredProxySettings.Mode == offMode:
...
case currProxySettings.Mode == defaultMode && desiredProxySettings.Mode == defaultMode && !hetznerInCurrent && !hetznerInDesired:
...
case currProxySettings.Mode == defaultMode && !hetznerInCurrent && desiredProxySettings.Mode == offMode:
...
case currProxySettings.Mode == offMode && desiredProxySettings.Mode == defaultMode && !hetznerInDesired:
...
default:
   return errors.New("failed to determine if proxy update is needed")
}
```